From 92f812ebf92a911d001ac6143f558d20ccce70a3 Mon Sep 17 00:00:00 2001
From: Peter Rekdal Khan-Sunde <peter.sunde@gmail.com>
Date: Thu, 6 Jun 2024 13:26:15 +0200
Subject: [PATCH 8/8] Merge pull request #3751 from peters:4.x

Add conditional include for <cuda/std/tuple> to support CUDA 12.4+ #3751

Fixes https://github.com/opencv/opencv_contrib/issues/3752

Adds a preprocessor check to conditionally include <cuda/std/tuple>  only for CUDA versions 12.4 and above. This ensures backward compatibility  with older legacy CUDA versions.

Related to pull request #3744

Upstream-Status: Backport [https://github.com/opencv/opencv_contrib/pull/3751]
Signed-off-by: Matt Madison <matt@madison.systems>

### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [x] I agree to contribute to the project under Apache 2 License.
- [x] To the best of my knowledge, the proposed patch is not based on a code under GPL or another license that is incompatible with OpenCV
- [x] The PR is proposed to the proper branch
- [ ] There is a reference to the original bug report and related work
- [ ] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [ ] The feature is well documented and sample code can be built with the project CMake
---
 modules/cudev/include/opencv2/cudev/ptr2d/zip.hpp | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/modules/cudev/include/opencv2/cudev/ptr2d/zip.hpp b/modules/cudev/include/opencv2/cudev/ptr2d/zip.hpp
index f5a3f8c8..f60ab5d0 100644
--- a/modules/cudev/include/opencv2/cudev/ptr2d/zip.hpp
+++ b/modules/cudev/include/opencv2/cudev/ptr2d/zip.hpp
@@ -49,7 +49,9 @@
 #include "../common.hpp"
 #include "../util/tuple.hpp"
 #include "traits.hpp"
+#if defined(__CUDACC_VER_MAJOR__) && (__CUDACC_VER_MAJOR__ > 12 || (__CUDACC_VER_MAJOR__ == 12 && __CUDACC_VER_MINOR__ >= 4))
 #include <cuda/std/tuple>
+#endif
 
 namespace cv { namespace cudev {
 
@@ -176,6 +178,7 @@ template <class PtrTuple> struct PtrTraits< ZipPtrSz<PtrTuple> > : PtrTraitsBase
 
 }}
 
+#if defined(__CUDACC_VER_MAJOR__) && (__CUDACC_VER_MAJOR__ > 12 || (__CUDACC_VER_MAJOR__ == 12 && __CUDACC_VER_MINOR__ >= 4))
 _LIBCUDACXX_BEGIN_NAMESPACE_STD
 
 template< class... Types >
@@ -198,3 +201,4 @@ struct tuple_element<N, cv::cudev::ZipPtrSz<tuple<Types...> > >
 _LIBCUDACXX_END_NAMESPACE_STD
 
 #endif
+#endif
-- 
2.43.0

